### PR TITLE
[WIP] Track element instances in running application

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Allows invalidation of a subset of the widget tree, when provided with file/line/column information. Currently registered in the WidgetInspector extensions due to the visibility constraints on `_Location`. Probably doesn't quite work yet

cc @jacob314

### CODE REVIEW

**Review by jacob314 — COMMENTED** [review]
I've got a local CL that does this but optimizes slightly by taking advantage that the _Location objects created by the VM are all const. Let me split it up from other hot ui prototype code on my Flutter branch and get it out for you to review. The other performance concern here is that deactivateElement call is O(elements created at the location). In my local prototype I'd used a Set<Element> but that isn't great either. 
One other idea is to add two debug fields on the Element class so that we can create a doubly linked list of Element objects tracking all the active Elements at each location. That way adding and removing elements can be O(1). I'm not sure how far we are from be able to treeshake unused debug fields from release builds but if that was the case then adding the debug fields to Element would definitely be the way to go.

**Comment by jonahwilliams** [comment]
Took another pass on this based on the double linked list approach. If we use expandos that are kept in the widget inspector class, then we don't need to add any fields to Element

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -3208,6 +3209,8 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
     _updateInheritance();
     assert(() {
       _debugLifecycleState = _ElementLifecycle.active;
+      WidgetInspectorService.instance.activateElement(this);
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter/lib/src/widgets/framework.dart]: Ahh, right - this code assumes that widget doesn't change. I'll tinker with that a bit.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 2]
```diff
@@ -708,6 +709,22 @@ mixin WidgetInspectorService {
...
+  // All active locations in a given file. There should only be a handful per
+  // file, making a linear search fast enough.
+  final Map<String, Set<_Location>> _locations = HashMap<String, Set<_Location>>();
+
+  // Fields used to track active instances of the element for faster
+  // invalidations. Together these form a double-linked list of elements,
+  // where every element corresponds to a Widget with an indentical location.
+  // The expando allows adding previous and next pointers to the element clas
+  // without modifying the public API. Unlike a map, it only weakly retains objects.
+  final Expando<Element> _previousElements = Expando<Element>();
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart]: nit: 
`_previousElement`, `_nextElement` as this is an expando for a single next element.

**Comment by jonahwilliams** [comment]
I'm going to leave this as a draft (or close it) for the short/medium term and pursue the approach of walking the entire element tree. Then I'll be able to say more easily whether this is a performance improvement, once we account for the edge cases

**Comment by jonahwilliams** [comment]
I'm going to close this one for now (for above reasons) to remove from PR queue

**Pull request closed without merging**